api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+	"c3.8xlarge":  60 * bytesize.GiB,
+	"g2.2xlarge":  15 * bytesize.GiB,
+	"g2.8xlarge":  60 * bytesize.GiB,
+	"x1.32xlarge": 1952 * bytesize.GiB,
+	"r3.large":    15.25 * bytesize.GiB,
+	"r3.xlarge":   30.5 * bytesize.GiB,
+	"r3.2xlarge":  61 * bytesize.GiB,
+	"r3.4xlarge":  122 * bytesize.GiB,
+	"r3.8xlarge":  244 * bytesize.GiB,
+	"i3.large":    15.25 * bytesize.GiB,
+	"i3.xlarge":   30.5 * bytesize.GiB,
+	"i3.2xlarge":  61 * bytesize.GiB,
+	"i3.4xlarge":  122 * bytesize.GiB,
+	"i3.8xlarge":  244 * bytesize.GiB,
+	"d2.xlarge":   30.5 * bytesize.GiB,
+	"d2.2xlarge":  61 * bytesize.GiB,
+	"d2.4xlarge":  122 * bytesize.GiB,
+	"d2.8xlarge":  244 * bytesize.GiB,
+}
 
 type EnvironmentScaler struct {
+	Client              *awsc.Client
+	EnvironmentProvider provider.EnvironmentProvider
+	ServiceProvider     provider.ServiceProvider
+	TaskProvider        provider.TaskProvider
+	JobStore            job.Store
+	Config              config.APIConfig
+	deployCache         map[string][]ResourceConsumer
 }
 
-func NewEnvironmentScaler() *EnvironmentScaler {
-	return &EnvironmentScaler{}
+const (
+	Memory = iota
+	CPU
+)
+
+func NewEnvironmentScaler(a *awsc.Client, e provider.EnvironmentProvider, s provider.ServiceProvider, t provider.TaskProvider, j job.Store, c config.APIConfig) *EnvironmentScaler {
+	return &EnvironmentScaler{
+		Client:              a,
+		EnvironmentProvider: e,
+		ServiceProvider:     s,
+		TaskProvider:        t,
+		JobStore:            j,
+		Config:              c,
+		// deployCache: map[string][]*Resource,
+	}
 }
 
 func (e *EnvironmentScaler) Scale(environmentID string) error {
-	return fmt.Errorf("EnvironmentScaler not implemented")
+	clusterName := addLayer0Prefix(e.Config.Instance(), environmentID)
+
+	resourceProviders, err := e.getResourceProviders(clusterName)
+	if err != nil {
+		return err
+	}
+
+	humanReadableProviders := ""
+	for _, r := range resourceProviders {
+		humanReadableProviders += fmt.Sprintf("\n    %#v", r)
+	}
+
+	log.Printf("[DEBUG] [EnvironmentScaler] resourceProviders for env '%s': %s", environmentID, humanReadableProviders)
+
+	var resourceConsumers []ResourceConsumer
+
+	// get pending service resource consumers
+	serviceConsumers, err := e.getResourceConsumers_PendingServices(clusterName)
+	if err != nil {
+		return err
+	}
+
+	resourceConsumers = append(resourceConsumers, serviceConsumers...)
+
+	// get pending task resource consumers in ECS
+	ecsTaskConsumers, err := e.getResourceConsumers_TasksInECS(clusterName)
+	if err != nil {
+		return err
+	}
+
+	resourceConsumers = append(resourceConsumers, ecsTaskConsumers...)
+
+	// get pending task resource consumers in jobs
+	jobTaskConsumers, err := e.getResourceConsumers_TasksInJobs(clusterName)
+	if err != nil {
+		return err
+	}
+
+	resourceConsumers = append(resourceConsumers, jobTaskConsumers...)
+
+	log.Printf("[DEBUG] [EnvironmentScaler] resourceConsumers for env '%s': %#v", environmentID, resourceConsumers)
+
+	var errs []error
```

Comment:
This looks like using multi errors for the sake of using multi errors. I'm not convinced this is a good idea. It seems this would open a door to a lot of unintended behavior when doing something like `scaleTo` when you know there were errors in the calculation.

Reply:
Overhauled error handling in this part of the code. ✅